### DISCUSSION TITLE
Welcome to JMeter JWT PreProcessor Discussions!

### DISCUSSION THREAD

**Discussion by sashikaR** [sashikaR/jwt-preProcessor#1] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with our users of this plugin. You can
  * Ask questions you’re wondering about.
  * Share ideas.
  * Add any issues, bugs or concerns that you have.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

**Comment by sit-md** [comment]
@sashikaR I am currently trying to use RSA256 to sign the token. Unfortunately whatever format I am trying, it is not accepting the private key. Can you point me into the right direction?

**Reply by sashikaR** [reply]
@sit-md Could you please share the Jmeter log? Please make sure to enable the INFO.

**Reply by sit-md** [reply]
So the key is enclosed in a
```
-----BEGIN RSA PRIVATE KEY----- 
-----END RSA PRIVATE KEY----- 
```
block and the key itself is Base64-encoded but not encrypted.

Here's the stack trace of the thrown exception:
```
2023-10-13 16:10:51,194 ERROR o.a.j.t.JMeterThread: Error while processing sampler: 'jp@gc - Dummy Sampler'.java.lang.IllegalArgumentException: Illegal base64 character 2d
	at java.util.Base64$Decoder.decode0(Base64.java:848) ~[?:?]
	at java.util.Base64$Decoder.decode(Base64.java:566) ~[?:?]
	at java.util.Base64$Decoder.decode(Base64.java:589) ~[?:?]
	at sr.jmeter.jwt.preprocessor.service.util.JwtUtil.convertPrivateKey(JwtUtil.java:35) ~[jmeter_plugins_jwt_pre_processor-1.0-SNAPSHOT.jar:?]
	at sr.jmeter.jwt.preprocessor.JwtServiceWrapper.generateJwt(JwtServiceWrapper.java:27) ~[jmeter_plugins_jwt_pre_processor-1.0-SNAPSHOT.jar:?]
	at sr.jmeter.jwt.preprocessor.JwtPreProcessor.process(JwtPreProcessor.java:43) ~[jmeter_plugins_jwt_pre_processor-1.0-SNAPSHOT.jar:?]
	at org.apache.jmeter.threads.JMeterThread.runPreProcessors(JMeterThread.java:983) ~[ApacheJMeter_core.jar:5.6.2]
	at org.apache.jmeter.threads.JMeterThread.executeSamplePackage(JMeterThread.java:561) ~[ApacheJMeter_core.jar:5.6.2]
	at org.apache.jmeter.threads.JMeterThread.processSampler(JMeterThread.java:501) [ApacheJMeter_core.jar:5.6.2]
	at org.apache.jmeter.threads.JMeterThread.run(JMeterThread.java:268) [ApacheJMeter_core.jar:5.6.2]
	at java.lang.Thread.run(Thread.java:833) [?:?]
```

I also tried only providing the key, without the begin and end block, leading to the same stack trace with a different character being blamed.

**Reply by sashikaR** [reply]
Hi @sit-md ,

I made the changes to support RSA PKCS#8. Could you please check with this release?
https://github.com/sashikaR/jwt-preProcessor/releases/download/v1.1.0/jmeter_plugins_jwt_pre_processor-1.1.0-SNAPSHOT.jar

You have to manually install the jar into your JMeter lib/ext folder.

**Reply by sashikaR** [reply]
Hi @sit-md ,

The plugin's new version(1.1.0) is now available in the Jmeter Plugin Manager. Would you be able to try with this version? Please remove the previous jar from your lib/ext location.

**Reply by sit-md** [reply]
Hi @sashikaR,

I installed the latest version through the Plugin Manager and it can now handle the PKCS8 format. However, when calling the endpoint that validates the JWT token, I receive an error from there because the data could not be verified. It is a .NET 6 function app and the code reads:
```
if (!key.VerifyData(encodedBytes, signature, HashAlgorithmName.SHA256, RSASignaturePadding.Pkcs1))
                {
                    throw new JsonWebTokenValidationException();
                }
```

and subsequently:
```
_rsa.VerifyData(data, signature, hashAlgorithm, padding)
```
whereas _rsa is an instance of the RSA class, which is a standard component in .NET 6.

Any ideas as to why that happens?

**Reply by sashikaR** [reply]
Hi @sit-md , Can you check the generated JWT with https://jwt.io/  ? This will be helpful to rectify the issue either in generated JWT or the service you are calling ?

**Reply by sit-md** [reply]
@sashikaR I actually did and the JWT token created is identical to a test instance that issues JWT tokens for testing purposes on non-production systems. Still it fails for the token created by the plugin but succeeds for the one issued by the test instance.

**Reply by sashikaR** [reply]
@sit-md , Then there should be some difference. Could you double check on the signing algorithms are identical in both JWT? It should be same in JWT header when extracted. Also double check and confirm that the private key that you have used for both tokens are same.

**Comment by RadhaUdhaykumar** [comment]
Hi @sashikaR , I tried installing the  JWT Pre-processor plugin 1.0.2 from the plugin manager. I am using Jmeter version 5.6.2 . Once the plugin gets installed and the Jmeter gets restarted, I am unable to see the option under Preprocessor. When I checked in the Plugin Manager under Installed Plugin , neither it shows up there. It still remains in the available plugins tab only. I am sure that I am clicking the "Apply changes and restart Jmeter" button at the botom of the Plugin Manager
